### PR TITLE
Adapter: promoted for structural similarity with DBAL

### PR DESCRIPTION
**Pull request by bryanlatten**
- BREAKING: promoting top-level elements to the src root
- Added integration tests, factory to 100%
- Removed 5.4, 5.5, hhvm travis testing (adapters not available everywhere)

### CODE REVIEW

File: tests/unit/FactoryTest.php [hunk 1]
```diff
@@ -10,6 +10,102 @@ class FactoryTest extends BaseTest {
...
+      ->will( $this->returnValue( false ) );
+
+    $factory->__construct();
+
+  } // constructUnavailable
+
+
+  /**
+   * @test
+   * @expectedException Behance\NBD\Cache\CacheException
```
Comment (be-dmitry) [inline tests/unit/FactoryTest.php]: Maybe get more specific with the exception type

File: src/Adapters/MemcacheAdapter.php [hunk 2]
```diff
@@ -37,10 +37,6 @@ class MemcacheAdapter extends AdapterAbstract {
    */
   public function __construct( EventDispatcherInterface $event_dispatcher = null, \Memcache $instance = null ) {
 
-    if ( !extension_loaded( 'memcache' ) ) {
-      throw new SystemRequirementException( 'Memcache extension is required' );
-    }
```
Comment (bryanlatten) [inline src/Adapters/MemcacheAdapter.php]: now checked in the factory

File: src/Adapters/MemcachedAdapter.php [hunk 3]
```diff
@@ -20,10 +20,6 @@ class MemcachedAdapter extends AdapterAbstract {
    */
   public function __construct( EventDispatcherInterface $event_dispatcher = null, \Memcached $instance = null ) {
 
-    if ( !extension_loaded( 'memcached' ) ) {
```
Comment (bryanlatten) [inline src/Adapters/MemcachedAdapter.php]: now checked in the factory

**Comment by DaveStein** [comment]
Since it's breaking change, I assume the release will be a major version jump

**Comment by DaveStein** [comment]
:+1:

**Comment by be-dmitry** [comment]
:+1:

**Pull request merged**